c compilation / install issue and weird message on bufenter / start R

Hi, I am really sorry to make an issue, but it was working on my main nvim config on macos, but now it isn't and I have tried all the usual things, different clang compiler, minimal config, different version etc but to no avail. 

Strangely i can install from the repo with 
```
[I] ~/r/p/R.nvim  rm -rf /Users/user/Library/R/arm64/4.4/library/nvimcom                                                                       main ‚¨á
                     R CMD INSTALL nvimcom
* installing to library '/Users/user/Library/R/arm64/4.4/library'
* installing *source* package 'nvimcom' ...
** using staged installation
** libs
using C compiler: 'Apple clang version 16.0.0 (clang-1600.0.26.6)'
using SDK: 'MacOSX15.2.sdk'
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -
c nvimcom.c -o nvimcom.o
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -
c rd2md.c -o rd2md.o
clang -arch arm64 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -L/Library/Frameworks/R.framework/Resources/lib -L/opt/R/arm64/
lib -o nvimcom.so common.o init.o nvimcom.o rd2md.o -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation
(cd apps; make)
make[1]: Nothing to be done for `all'.
installing via 'install.libs.R' to /Users/user/Library/R/arm64/4.4/library/00LOCK-nvimcom/00new/nvimcom
** R
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (nvimcom)
```

But not when inside nvim (printout from RDebugInfo)
```
before_rns.R stdout: * checking for file '/Users/user/repos/public/R.nvim/nvimcom/DESCRIPTION' ...
 OK
* preparing 'nvimcom':
* checking DESCRIPTION meta-information ...
 OK
* cleaning src
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
* building 'nvimcom_0.9.64.tar.gz'
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c common.c -o common.o
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c init.c -o init.o
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c nvimcom.c -o nvimcom.o
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c rd2md.c -o rd2md.o
clang -arch arm64 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -L/Library/Frameworks/R.framework/Resources/lib -L/opt/R/arm64/lib -o nvimcom.so common.o init.o nvimcom.o rd2md.o -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation
(cd apps; make)
make[1]: Nothing to be done for `all'.
Error: package or namespace load failed for 'nvimcom' in dyn.load(file, DLLpath = DLLpath, ...): unable to load shared object '/Users/user/Library/R/arm64/4.4/library/00LOCK-nvimcom/00new/nvimcom/libs/nvimcom.so':  dlopen(/Users/user/Library/R/arm64/4.4/library/00LOCK-nvimcom/00new/nvimcom/libs/nvimcom.so, 0x0006): symbol not found in flat namespace '_fmt_txt'Error: loading failedExecution halted

Time:
  check health (async): 0.006841333
  global setup: 0.048954833
  before_rns.R (async): 2.442556958
  pdf setup (async): 0.00226825
  unix setup: 0.011611792
  rmd setup: 0.001482916
before_rns.R stderr: 
.build_packages() exit status 0
* installing to library '/Users/user/Library/R/arm64/4.4/library'
* installing *source* package 'nvimcom' ...
** using staged installation
** libs
using C compiler: 'Apple clang version 16.0.0 (clang-1600.0.26.6)'
using SDK: 'MacOSX15.2.sdk'
installing via 'install.libs.R' to /Users/user/Library/R/arm64/4.4/library/00LOCK-nvimcom/00new/nvimcom
** R
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** testing if installed package can be loaded from temporary location
ERROR: loading failed
* removing '/Users/user/Library/R/arm64/4.4/library/nvimcom'
```

This is my config:
```
-- Must be before creating other maps:
vim.g.mapleader = ' '
vim.g.maplocalleader = ' '


-- Set your global variables and options above this line --

local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({
    "git",
    "clone",
    "--filter=blob:none",
    "https://github.com/folke/lazy.nvim.git",
    "--branch=stable", -- latest stable release
    lazypath,
  })
end
vim.opt.rtp:prepend(lazypath)

require("lazy").setup({
  -- "R-nvim/R.nvim",
	{dir = '/Users/user/repos/public/R.nvim'},
  "R-nvim/cmp-r",
  {
    "nvim-treesitter/nvim-treesitter",
    run = ":TSUpdate",
    config = function ()
      require("nvim-treesitter.configs").setup({
	sync_install = true,
	ensure_installed = {
	  "r",
	  "markdown",
	  "markdown_inline",
	  "rnoweb",
          "yaml",
					"latex",
	},
	highlight = {
	  enable = true,
	},
      })
    end,
  },

  {
    "hrsh7th/nvim-cmp",
    config = function()
      local cmp = require("cmp")
      cmp.setup({
	sources = {{ name = "cmp_r" }},
	mapping = cmp.mapping.preset.insert({
	  ['<CR>'] = cmp.mapping.confirm({ select = false }),
	  -- During auto-completion, press <Tab> to select the next item.
	  ['<Tab>'] = cmp.mapping(function(fallback)
	    if cmp.visible() then
	      cmp.select_next_item({ behavior = cmp.SelectBehavior.Insert })
	    elseif has_words_before() then
	      cmp.complete()
	    else
	      fallback()
	    end
	  end, { 'i', 's' }),
	  ['<S-Tab>'] = cmp.mapping(function(fallback)
	    if cmp.visible() then
	      cmp.select_prev_item({ behavior = cmp.SelectBehavior.Insert })
	    else
	      fallback()
	    end
	  end, { 'i', 's' }),
	}),
      })
      require("cmp_r").setup({ })
    end,
  },
}, {})

-- vim: sw=2 noexpandtab
```
where the local R.nvim is arbitrarily at commit 3cbab46

And i tried `iconv -f UTF-8 -t ASCII//TRANSLIT nvimcom/src/rd2md.c > nvimcom/src/rd2md.c`
 and `iconv -f UTF-8 -t ASCII//TRANSLIT nvimcom/src/nvimcom.c.c > nvimcom/src/nvimcom.c`that didnt work either.

Really stuck here sorry!

Yes that is my username
Also, this is the original error i was getting, and the reason why i tried to fix the bug with iconv
```
Why build nvimcom: Nvimcom not installed
before_rns.R stderr: Warning in readLines(ff, warn = FALSE) :  invalid input found on input connection 'nvimcom/src/nvimcom.c'
Warning in readLines(ff, warn = FALSE) :  invalid input found on input connection 'nvimcom/src/rd2md.c'

.build_packages() exit status 0
* installing to library '/Users/user/Library/R/arm64/4.4/library'
* installing *source* package 'nvimcom' ...
** using staged installation
** libs
using C compiler: 'Apple clang version 16.0.0 (clang-1600.0.26.6)'
using SDK: 'MacOSX15.2.sdk'
nvimcom.c:554:28: warning: missing terminating '"' character [-Winvalid-pp-token]  554
 |         snprintf(buf, 127, "      |                            ^nvimcom.c:554:9: error: unterminated function-like macro invocation  554 |         
snprintf(buf, 127, "      
|         ^/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/secure/_stdio.h:56:
9: note: macro 'snprintf' defined here   56 | #de
fine snprint
f(str, 
len, ...
) \      | 
        ^
nvimcom.c:554
:29: error: expected '}'  554 |   
      snpri
ntf(bu
f, 12
7, "
      |                             ^nvimcom.c:553:22:
 note: to match this '{'  553 |     
if (xgroup == 2) {      |                      ^nvimcom.c:
554:29: error: expected '}'
  554 |   
      
snprin
tf(buf,
 127,
 "      | 
                            ^nvimcom.c:465:53: note: to match this '{'  465 |   
            
        
    
    
  char
 *p, i
nt dep
th) {
      |                                                     ^
1 warning and 3
 errors generated.
make: *** [nvimcom.o] Error 1
ERROR: compilation failed for package 'nvimcom'
* removing '/Users/user/Library/R/arm64/4.4/library/nvimcom'
Error in do_exit(status = status) : .install_packages() exit status 1
Error in do_exit(status = status) : .install_packages() exit status 1
Execution halted

before_rns.R stdout: * checking for file '/Users/user/.local/share/quarto-nvim-kickstarter/lazy/R.nvim/nvimcom/DESCRIPTION' ...
 OK
* preparing 'nvimcom':
* checking DESCRIPTION meta-information ...
 OK
* cleaning src
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
* building 'nvimcom_0.9.66.tar.gz'
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c common.c -o common.o
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c init.c -o init.o
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c nvimcom.c -o nvimcom.o

Time:
  before_rns.R (async): 0.941039042
  check health (async): 0.004446792
  global setup: 0.012644709
  pdf setup (async): 0.000581792
  unix setup: 0.005420458
  rmd setup: 0.000819042
```

Sorry, that didn't work again. I genuinely have no idea here. Other than something updated in 15.3.2... @richardsc 

Just to confirm i had that branch here is lazy package manager's printout:
```
    ‚óè R.nvim 47.45ms Ó´ì start

        dir    /Users/user/.local/share/quarto-nvim-kickstarter/lazy/R.nvim

        url    https://github.com/R-nvim/R.nvim

        branch ascii_nvimcom

        commit fc0b423

        readme README.md

        help   |R.nvim.txt|

      ‚úî [task] clone 1139.29ms

        Cloning into '/Users/user/.local/share/quarto-nvim-kickstarter/lazy/R.nvim'...

        remote: Enumerating objects: 2094, done.        

        remote: Counting objects: 100% (828/828), done.        

        remote: Compressing objects: 100% (186/186), done.        

        remote: Total 2094 (delta 726), reused 653 (delta 642), pack-reused 1266 (from 2)        

        Receiving objects: 100% (2094/2094), 343.26 KiB | 5.72 MiB/s, done.

        Resolving deltas: 100% (1006/1006), done.

      ‚úî [task] checkout 0.55ms

      ‚úî [task] docs 4.01ms
```

RDebugInfo
```
Why build nvimcom: Nvimcom not installed
before_rns.R stderr: Warning in readLines(ff, warn = FALSE) :  invalid input found on input connection 'nvimcom/src/rd2md.c'

.build_packages() exit status 0
* installing to library '/Users/user/Library/R/arm64/4.4/library'
* installing *source* package 'nvimcom' ...
** using staged installation
** libs
using C compiler: 'Apple clang version 16.0.0 (clang-1600.0.26.6)'
using SDK: 'MacOSX15.2.sdk'
rd2md.c:42:21: warning: missing terminating '"' character [-Winvalid-pp-token]
   42 |     {"item{", 5, 9, "\x14        |                     ^rd2md.c:42:21: error: expected expression
rd2md.c:42:28: error: expected '}'   42
 |     {"item{", 5
, 9, "\x14
  
      |                            ^rd2md.c:42:
5: note: to match this '{'   42 |  
   {"item{", 5, 9,
 "\x14
        | 
    ^rd2md.c:
42:28: error: expected '}'   42 |    
 {"item{", 5
, 9, "\
x14
        | 
                           ^rd2md.c:31:30
: note: to match this '{'   31 | sta
tic struc
t pa
tter
n rd[
] = {      
|                              ^
rd2md.c:42:28: error: expected ';' after top level declarator
   42 |     {"it
em{", 5
, 9, "
\x14  
      |                            ^      |                            ;

1 warning and 4
 errors generated.
make: *** [rd2md.o] Error 1
ERROR: compilation failed for package 'nvimcom'
* removing '/Users/user/Library/R/arm64/4.4/library/nvimcom'
Error in do_exit(status = status) : .install_packages() exit status 1
Error in do_exit(status = status) : .install_packages() exit status 1Execution halted

Time:
  global setup: 0.09221575
  before_rns.R (async): 1.309360958
  pdf setup (async): 0.020611
  check health (async): 0.004953625
  rmd setup: 0.001038833
  unix setup: 0.006914875
before_rns.R stdout: * checking for file '/Users/user/.local/share/quarto-nvim-kickstarter/lazy/R.nvim/nvimcom/DESCRIPTION' ...
 OK
* preparing 'nvimcom':
* checking DESCRIPTION meta-information ...
 OK
* cleaning src
* checking for LF line-endings in source and make files and shell scripts
* checking for empty or unneeded directories
* building 'nvimcom_0.9.67.tar.gz'
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c common.c -o common.o
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c init.c -o init.o
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c nvimcom.c -o nvimcom.o
clang -arch arm64 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG   -I/opt/R/arm64/include    -fPIC  -falign-functions=64 -Wall -g -O2  -c rd2md.c -o rd2md.o
```

--- a/nvimcom/DESCRIPTION
+++ b/nvimcom/DESCRIPTION
@@ -1,6 +1,6 @@
 Package: nvimcom
-Version: 0.9.66
-Date: 2025-03-25
+Version: 0.9.67
+Date: 2025-03-27
 Title: Intermediate the Communication Between R and Neovim
 Author: Jakson Aquino
 Maintainer: Jakson Alves de Aquino <jalvesaq@gmail.com>

--- a/nvimcom/src/nvimcom.c
+++ b/nvimcom/src/nvimcom.c
@@ -551,10 +551,11 @@ static char *nvimcom_glbnv_line(SEXP *x, const char *xname, const char *curenv,
 
     // Add the object length
     if (xgroup == 2) {
-        snprintf(buf, 127, " [%d, %d]", length(Rf_GetRowNames(*x)), length(*x));
+        snprintf(buf, 127, "\xc2\xa0[%d, %d]", length(Rf_GetRowNames(*x)),
+                 length(*x));
         p = str_cat(p, buf);
     } else if (xgroup == 3) {
-        snprintf(buf, 127, " [%d]", length(*x));
+        snprintf(buf, 127, "\xc2\xa0[%d]", length(*x));
         p = str_cat(p, buf);
     } else if (xgroup == 4) {
         SEXP cmdSexp, cmdexpr;
@@ -579,7 +580,7 @@ static char *nvimcom_glbnv_line(SEXP *x, const char *xname, const char *curenv,
                      curenv, xname);
         }
         UNPROTECT(2);
-        snprintf(buf, 127, " [%d]", len);
+        snprintf(buf, 127, "\xc2\xa0[%d]", len);
         p = str_cat(p, buf);
     }
 
@@ -997,12 +998,11 @@ static void SrcrefInfo(void) {
             size_t slen = strlen(CHAR(STRING_ELT(filename, 0)));
             char *buf = calloc(sizeof(char), (2 * slen + 56));
             char *buf2 = calloc(sizeof(char), (2 * slen + 56));
-            snprintf(buf, 2 * slen + 1, "%s",
-                    CHAR(STRING_ELT(filename, 0)));
+            snprintf(buf, 2 * slen + 1, "%s", CHAR(STRING_ELT(filename, 0)));
             nvimcom_squo(buf, buf2, 2 * slen + 32);
             snprintf(buf, 2 * slen + 55,
-                    "lua require('r.debug').jump('%s', %d)", buf2,
-                    asInteger(R_Srcref));
+                     "lua require('r.debug').jump('%s', %d)", buf2,
+                     asInteger(R_Srcref));
             send_to_nvim(buf);
             free(buf);
             free(buf2);

--- a/nvimcom/src/rd2md.c
+++ b/nvimcom/src/rd2md.c
@@ -18,7 +18,7 @@ typedef struct pattern {
    3: \cmd{s1}{s2}                  -> <s2>
    4: \cmd{s}{}                     -> ""
    5: \cmd[optional-argument]{s}    -> <s>
-   6: \href{s1}{s2}                 -> ‘s2’ <s1>
+   6: \href{s1}{s2}                 -> 's2' <s1> // curly single quotes
    7: \ifelse{{html|latex}{s1}{s2}} -> s2
    8: \code{\link{s}}               -> <s>
    9: \item{s1}{s2}                 -> `s1`: s2 || \n - s1
@@ -28,32 +28,35 @@ typedef struct pattern {
 // Optimized by frequency in titles and description of objects in default R
 // packages : \code (1814), \R (220), \emph (97), sQuote (76), \eqn (59),
 // \dQuote (41), \link (30), \pkg (29) etc...
+// We use hexadecimal codes to print Unicode characters because some
+// compilers report invalid code if we use non-ascii strings (see
+// https://github.com/R-nvim/R.nvim/issues/378#issuecomment-2758009615).
 static struct pattern rd[] = {
     // Order optimized for default R 4.2.2 packages
     {"code", 4, 8, "`", "`"},
     {"R", 1, 0, "*R*", "\000"},
     {"emph", 4, 1, "*", "*"},
     {"eqn", 3, 1, "*", "*"},
-    {"sQuote", 6, 1, "‘", "’"},
-    {"dQuote", 6, 1, "“", "”"},
+    {"sQuote", 6, 1, "\xe2\x80\x98", "\xe2\x80\x99"},
+    {"dQuote", 6, 1, "\xe2\x80\x9c", "\xe2\x80\x9d"},
     {"pkg", 3, 1, "\000", "\000"},
     {"linkS4class", 11, 1, "\000", "\000"},
-    {"link", 4, 5, "‘", "’"},
-    {"item{", 5, 9, "\x14  • ", "\x14"},
-    {"item ", 5, 0, "\x14  • ", "\000"},
+    {"link", 4, 5, "\xe2\x80\x98", "\xe2\x80\x99"},
+    {"item{", 5, 9, "\x14  \xe2\x80\xa2 ", "\x14"},
+    {"item ", 5, 0, "\x14  \xe2\x80\xa2 ", "\000"},
     {"itemize", 7, 1, "\x14", "\x14"},
-    {"item", 4, 1, "\x14  • ", "\000"},
+    {"item", 4, 1, "\x14  \xe2\x80\xa2 ", "\000"},
     {"dots", 4, 0, "...", "\000"},
     {"bold", 4, 1, "**", "**"},
-    {"file", 4, 1, "‘", "’"},
+    {"file", 4, 1, "\xe2\x80\x98", "\xe2\x80\x99"},
     {"option", 6, 1, "\000", "\000"},
     {"command", 7, 1, "`", "`"},
-    {"mu", 2, 0, "μ", "\000"},
+    {"mu", 2, 0, "\xce\xbc", "\000"},
     {"ifelse", 6, 7, NULL, NULL},
     {"samp", 4, 1, "`", "`"},
     {"env", 3, 1, "\000", "\000"},
     {"describe", 8, 1, "\x14", "\x14"},
-    {"Sigma", 5, 0, "Σ", "\000"},
+    {"Sigma", 5, 0, "\xce\xa3", "\000"},
 
     // Common in some other packages
     {"if{html}", 8, 4, NULL, NULL},
@@ -62,40 +65,40 @@ static struct pattern rd[] = {
     {"preformatted", 12, 1, "\x14```\x14", "```\x14"},
 
     // The rest
-    {"alpha", 5, 0, "α", "\000"},
-    {"beta", 4, 0, "β", "\000"},
-    {"Delta", 5, 0, "Δ", "\000"},
-    {"delta", 5, 0, "δ", "\000"},
-    {"epsilon", 7, 0, "ε", "\000"},
-    {"zeta", 4, 0, "ζ", "\000"},
-    {"theta", 5, 0, "θ", "\000"},
-    {"iota", 4, 0, "ι", "\000"},
-    {"kappa", 5, 0, "κ", "\000"},
-    {"eta", 3, 0, "η", "\000"},
-    {"gamma", 5, 0, "γ", "\000"},
-    {"lambda", 6, 0, "λ", "\000"},
-    {"nu", 2, 0, "ν", "\000"},
-    {"xi", 2, 0, "ξ", "\000"},
-    {"omega", 5, 0, "ω", "\000"},
-    {"Omega", 5, 0, "Ω", "\000"},
-    {"pi", 2, 0, "π", "\000"},
-    {"phi", 3, 0, "φ", "\000"},
-    {"chi", 3, 0, "χ", "\000"},
-    {"psi", 3, 0, "ψ", "\000"},
-    {"tau", 3, 0, "τ", "\000"},
-    {"upsilon", 7, 0, "υ", "\000"},
-    {"rho", 3, 0, "ρ", "\000"},
-    {"sigma", 5, 0, "σ", "\000"},
+    {"alpha", 5, 0, "\xce\xb1", "\000"},
+    {"beta", 4, 0, "\xce\xb2", "\000"},
+    {"Delta", 5, 0, "\xce\x94", "\000"},
+    {"delta", 5, 0, "\xce\xb4", "\000"},
+    {"epsilon", 7, 0, "\xce\xb5", "\000"},
+    {"zeta", 4, 0, "\xce\xb6", "\000"},
+    {"theta", 5, 0, "\xce\xb8", "\000"},
+    {"iota", 4, 0, "\xce\xb9", "\000"},
+    {"kappa", 5, 0, "\xce\xba", "\000"},
+    {"eta", 3, 0, "\xce\xb7", "\000"},
+    {"gamma", 5, 0, "\xce\xb3", "\000"},
+    {"lambda", 6, 0, "\xce\xbb", "\000"},
+    {"nu", 2, 0, "\xce\xbd", "\000"},
+    {"xi", 2, 0, "\xce\xbe", "\000"},
+    {"omega", 5, 0, "\xcf\x89", "\000"},
+    {"Omega", 5, 0, "\xce\xa9", "\000"},
+    {"pi", 2, 0, "\xcf\x80", "\000"},
+    {"phi", 3, 0, "\xcf\x86", "\000"},
+    {"chi", 3, 0, "\xcf\x87", "\000"},
+    {"psi", 3, 0, "\xcf\x88", "\000"},
+    {"tau", 3, 0, "\xcf\x84", "\000"},
+    {"upsilon", 7, 0, "\xcf\x85", "\000"},
+    {"rho", 3, 0, "\xcf\x81", "\000"},
+    {"sigma", 5, 0, "\xcf\x83", "\000"},
     {"log", 3, 0, "log", "\000"},
-    {"le", 2, 0, "≤", "\000"},
-    {"ge", 2, 0, "≥", "\000"},
-    {"ll", 2, 0, "≪", "\000"},
-    {"gg", 2, 0, "≫", "\000"},
-    {"infty", 5, 0, "∞", "\000"},
+    {"le", 2, 0, "\xe2\x89\xa4", "\000"},
+    {"ge", 2, 0, "\xe2\x89\xa5", "\000"},
+    {"ll", 2, 0, "\xe2\x89\xaa", "\000"},
+    {"gg", 2, 0, "\xe2\x89\xab", "\000"},
+    {"infty", 5, 0, "\xe2\x88\x9e", "\000"},
     {"tabular", 7, 3, "\x14", "\x14"},
     {"tab", 3, 0, "\t", "\000"},
     {"cr", 2, 0, "\x14", "\000"},
-    {"sqrt", 4, 1, "*√", "*"},
+    {"sqrt", 4, 1, "*\xe2\x88\x9a", "*"},
     {"strong", 6, 1, "**", "**"},
     {"email", 5, 1, "\000", "\000"},
     {"acronym", 7, 1, "\000", "\000"},
@@ -104,7 +107,7 @@ static struct pattern rd[] = {
     {"deqn", 4, 1, "*", "*"},
     {"cite", 4, 1, "\000", "\000"},
     {"url", 3, 1, "\000", "\000"},
-    {"ldots", 5, 0, "…", "\000"},
+    {"ldots", 5, 0, "\xe2\x80\xa6", "\000"},
     {"verb", 4, 1, "`", "`"},
     {"out", 3, 1, "\000", "\000"},
     {"examples", 8, 1, "\x14```r\x14", "```\x14"},
@@ -235,7 +238,7 @@ static void rd_md(char **o1, char **o2) {
                 *o1 = p1;
                 *o2 = p2;
                 return;
-            case 6: // \href{string1}{string2} -> ‘string2’ <string1>
+            case 6: // \href{string1}{string2} -> 'string2' <string1>
                 p2 += rd[i].len + 1;
                 p2a = p2;
                 p2a = p2a + find_matching_bracket(p2a);
@@ -243,9 +246,9 @@ static void rd_md(char **o1, char **o2) {
                 p2b = p2a + 2;
                 p3 = p2b + find_matching_bracket(p2b);
                 *p3 = 0;
-                p1 = insert_str(p1, "‘");
+                p1 = insert_str(p1, "\xe2\x80\x98");
                 p1 = insert_str(p1, p2b);
-                p1 = insert_str(p1, "’ <");
+                p1 = insert_str(p1, "\xe2\x80\x99 <");
                 p1 = insert_str(p1, p2);
                 p1 = insert_str(p1, ">");
                 p2 = p3 + 1;
@@ -306,7 +309,7 @@ static void rd_md(char **o1, char **o2) {
                 *o1 = p1;
                 *o2 = p2;
                 return;
-            case 9: // \item{string1}{string2} -> ‘string1’: string2
+            case 9: // \item{string1}{string2} -> 'string1': string2
                 p2 += rd[i].len;
                 p2a = p2;
                 p2a = p2a + find_matching_bracket(p2a);

